lib/services/cloud-service.ts

```diff
@@ -1,89 +1,108 @@
+	constructor(protected $errors: IErrors,
+		protected $fs: IFileSystem,
 		protected $httpClient: Server.IHttpClient,
-		protected $logger: ILogger) {
+		protected $logger: ILogger,
+		private $nsCloudOperationFactory: ICloudOperationFactory,
+		private $nsCloudS3Service: IS3Service,
+		private $nsCloudOutputFilter: ICloudOutputFilter,
+		private $processService: IProcessService) {
 		super();
+		this.cloudOperations = Object.create(null);
+		this.$processService.attachToProcessExitSignals(this, this.cleanup.bind(this));
 	}
 
-	protected async getObjectFromS3File<T>(pathToFile: string): Promise<T> {
-		return JSON.parse(await this.getContentOfS3File(pathToFile));
+	public abstract getServerOperationOutputDirectory(options: IOutputDirectoryOptions): string;
+
+	public async sendCloudMessage<T>(message: ICloudOperationMessage<T>): Promise<void> {
+		const cloudOperation = this.cloudOperations[message.cloudOperationId];
+		if (!cloudOperation) {
+			this.$errors.failWithoutHelp(`Cloud operation with id: ${message.cloudOperationId} not found.`);
+		}
+
+		await cloudOperation.sendMessage(message);
 	}
 
-	protected async getContentOfS3File(pathToFile: string): Promise<string> {
-		return (await this.$httpClient.httpRequest(pathToFile)).body;
+	protected abstract getServerResults(serverResult: ICloudOperationResult): IServerItem[];
+
+	protected async executeCloudOperation<T>(cloudOperationName: string, action: (cloudOperationId: string) => Promise<T>): Promise<T> {
+		const cloudOperationId: string = v4();
+		try {
+			this.$logger.info(`Starting ${cloudOperationName}. Cloud operation id: ${cloudOperationId}`);
+			const result = await action(cloudOperationId);
+
+			return result;
+		} catch (err) {
+			err.cloudOperationId = cloudOperationId;
+			throw err;
+		}
 	}
 
-	protected async waitForServerOperationToFinish(operationId: string, serverInformation: IServerResponse): Promise<void> {
-		const promiseRetryOptions = {
-			retries: CloudService.OPERATION_STATUS_CHECK_RETRY_COUNT,
-			minTimeout: CloudService.OPERATION_STATUS_CHECK_INTERVAL
-		};
-		return promiseRetry((retry, attempt) => {
-			return new Promise<IServerStatus>(async (resolve, reject) => {
-				try {
-					resolve(await this.getObjectFromS3File<IServerStatus>(serverInformation.statusUrl));
-				} catch (err) {
-					this.$logger.trace(err);
-					reject(new Error(this.failedToStartError));
+	protected async waitForServerOperationToFinish(cloudOperationId: string, serverResponse: IServerResponse, options: ICloudOperationExecutionOptions): Promise<ICloudOperationResult> {
+		const cloudOperationVersion = serverResponse.cloudOperationVersion || CloudService.CLOUD_OPERATION_VERSION_1;
+		const cloudOperation: ICloudOperation = this.$nsCloudOperationFactory.create(cloudOperationVersion, cloudOperationId, serverResponse);
+		this.cloudOperations[cloudOperationId] = cloudOperation;
+
+		cloudOperation.on(CloudCommunicationEvents.MESSAGE, (m: ICloudOperationMessage<any>) => {
```

Comment:
This handler is never removed

Reply:
1. I will remove all handlers in the cloud operation cleanup method.
2. We should always attach the handler because we are going to receive messages other than output and we need to proxy them. For example the inputRequest message.